### ISSUE TITLE
Error: Your environment contains multiple teams, but PPOTrainer doesn't support adversarial games. Enable self-play to train adversarial games.

### ISSUE THREAD

**Issue by nntsrb** [Unity-Technologies/ml-agents#4199]
I have recieved this error
eventhough I had added this line in the config file

    self_play:
      window: 10
      play_against_latest_model_ratio: 0.5
      save_steps: 50000
      swap_steps: 2000
      team_change: 100000

what exactly mean by enable this function ?

**Comment by andrewcoh** [comment]
Hi @nntsrb 

Did you set your team ids in the behavior parameters script? Also, can you share your full config file and the full output from the terminal?

**Comment by nntsrb** [comment]
behaviors:
  TileBase:
    trainer_type: ppo
    hyperparameters:
      batch_size: 2048
      buffer_size: 20480
      learning_rate: 0.0003
      beta: 0.005
      epsilon: 0.2
      lambd: 0.95
      num_epoch: 3
      learning_rate_schedule: constant
    network_settings:
      normalize: false
      hidden_units: 512
      num_layers: 2
      vis_encode_type: simple
    reward_signals:
      extrinsic:
        gamma: 0.99
        strength: 1.0
    keep_checkpoints: 5
    max_steps: 50000000
    time_horizon: 1000
    summary_freq: 10000
    threaded: true
    self_play:
      save_steps: 50000
      team_change: 200000
      swap_steps: 2000
      window: 10
      play_against_latest_model_ratio: 0.5
      initial_elo: 1200.0
curriculum:
  TileBase:
    measure: reward
    thresholds: [-1.0, -0.7, -0.3, 0.3, 0.7, 1.0]
    min_lesson_length: 80
    signal_smoothing: true
    parameters:
       enemy_count: [5.0, 4.0, 3.0, 3.0, 2.0, 2.0, 2.0]

yes I have set team ids in bhv parameters 0 and 1

**Comment by nntsrb** [comment]
2020-07-08 21:05:33.264793: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'cudart64_101.dll'; dlerror: cudart64_101.dll not found
2020-07-08 21:05:33.269573: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
WARNING:tensorflow:From c:\users\44750\desktop\vrtl-env\python-envs\sample-env\lib\site-packages\tensorflow\python\compat\v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
Instructions for updating:
non-resource variables are not supported in the long term
2020-07-08 21:05:34 INFO [environment.py:199] Listening on port 5004. Start training by pressing the Play button in the Unity Editor.
2020-07-08 21:05:38 INFO [environment.py:108] Connected to Unity environment with package version 1.1.0-preview and communication version 1.0.0
2020-07-08 21:05:38 INFO [environment.py:265] Connected new brain:
TileBaseISO?team=1
2020-07-08 21:05:38 INFO [environment.py:265] Connected new brain:
TileBaseISO?team=0
2020-07-08 21:05:38 WARNING [trainer_util.py:96] Metacurriculum enabled, but no curriculum for brain TileBaseISO. Brains with curricula: dict_keys(['TileBase']).
2020-07-08 21:05:38.568555: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2020-07-08 21:05:38.577295: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x20e71ddfb10 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-07-08 21:05:38.580400: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-07-08 21:05:38.589405: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library nvcuda.dll
2020-07-08 21:05:39.094596: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.71GHz coreCount: 46 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.23GiB/s
2020-07-08 21:05:39.100176: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 1 with properties:
pciBusID: 0000:02:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.71GHz coreCount: 46 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.23GiB/s
2020-07-08 21:05:39.105756: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'cudart64_101.dll'; dlerror: cudart64_101.dll not found
2020-07-08 21:05:39.111379: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_10.dll
2020-07-08 21:05:39.118338: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_10.dll
2020-07-08 21:05:39.121647: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_10.dll
2020-07-08 21:05:39.128544: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_10.dll
2020-07-08 21:05:39.132443: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_10.dll
2020-07-08 21:05:39.139373: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2020-07-08 21:05:39.141850: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2020-07-08 21:05:39.149195: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-07-08 21:05:39.153181: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      0 1
2020-07-08 21:05:39.154718: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 0:   N Y
2020-07-08 21:05:39.156062: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 1:   Y N
2020-07-08 21:05:39.159928: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x20e721397a0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2020-07-08 21:05:39.163508: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): GeForce RTX 2080, Compute Capability 7.5
2020-07-08 21:05:39.166463: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (1): GeForce RTX 2080, Compute Capability 7.5
2020-07-08 21:05:39 WARNING [stats.py:197] events.out.tfevents.1594238628.DESKTOP-VB7B3QL was left over from a previous run. Deleting.
2020-07-08 21:05:39 INFO [stats.py:130] Hyperparameters for behavior name TileBaseISO:
        trainer_type:   ppo
        hyperparameters:
          batch_size:   1024
          buffer_size:  10240
          learning_rate:        0.0003
          beta: 0.005
          epsilon:      0.2
          lambd:        0.95
          num_epoch:    3
          learning_rate_schedule:       linear
        network_settings:
          normalize:    False
          hidden_units: 128
          num_layers:   2
          vis_encode_type:      simple
          memory:       None
        reward_signals:
          extrinsic:
            gamma:      0.99
            strength:   1.0
        init_path:      None
        keep_checkpoints:       5
        checkpoint_interval:    500000
        max_steps:      500000
        time_horizon:   64
        summary_freq:   50000
        threaded:       True
        self_play:      None
        behavioral_cloning:     None
2020-07-08 21:05:39.184578: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.71GHz coreCount: 46 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.23GiB/s
2020-07-08 21:05:39.189616: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 1 with properties:
pciBusID: 0000:02:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.71GHz coreCount: 46 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.23GiB/s
2020-07-08 21:05:39.195256: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'cudart64_101.dll'; dlerror: cudart64_101.dll not found
2020-07-08 21:05:39.197638: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_10.dll
2020-07-08 21:05:39.199991: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_10.dll
2020-07-08 21:05:39.202936: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_10.dll
2020-07-08 21:05:39.204870: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_10.dll
2020-07-08 21:05:39.206848: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_10.dll
2020-07-08 21:05:39.208753: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2020-07-08 21:05:39.210825: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2020-07-08 21:05:39.218781: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-07-08 21:05:39.222538: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      0 1
2020-07-08 21:05:39.223943: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 0:   N Y
2020-07-08 21:05:39.225291: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 1:   Y N
2020-07-08 21:05:40.058013: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.71GHz coreCount: 46 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.23GiB/s
2020-07-08 21:05:40.063121: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 1 with properties:
pciBusID: 0000:02:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.71GHz coreCount: 46 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 417.23GiB/s
2020-07-08 21:05:40.071023: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'cudart64_101.dll'; dlerror: cudart64_101.dll not found
2020-07-08 21:05:40.085267: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_10.dll
2020-07-08 21:05:40.087924: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_10.dll
2020-07-08 21:05:40.099174: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_10.dll
2020-07-08 21:05:40.101556: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_10.dll
2020-07-08 21:05:40.103496: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_10.dll
2020-07-08 21:05:40.117907: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2020-07-08 21:05:40.120753: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2020-07-08 21:05:40.139303: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-07-08 21:05:40.143479: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      0 1
2020-07-08 21:05:40.145738: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 0:   N Y
2020-07-08 21:05:40.147205: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 1:   Y N
2020-07-08 21:05:40 WARNING [trainer.py:226] Your environment contains multiple teams, but PPOTrainer doesn't support adversarial games. Enable self-play to                     train adversarial games.

**Comment by andrewcoh** [comment]
Looks like the behavior name in your yaml doesn't match the behavior name in the behavior parameters script. The yaml is looking for `TileBase` but the behavior parameters script has `TileBaseISO`

**Comment by nntsrb** [comment]
> Looks like the behavior name in your yaml doesn't match the behavior name in the behavior parameters script. The yaml is looking for `TileBase` but the behavior parameters script has `TileBaseISO`

ohhh yes exactly ! thank you

**Comment by andrewcoh** [comment]
No problem. I'm closing this issue. Please open another if you encounter any other problems.
